include/boost/multiprecision/traits/max_digits10.hpp

```diff
@@ -0,0 +1,59 @@
+///////////////////////////////////////////////////////////////
+//  Copyright 2012 John Maddock. Distributed under the Boost
+//  Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at https://www.boost.org/LICENSE_1_0.txt
+
+#ifndef BOOST_MATH_MAX_DIGITS10_HPP
+#define BOOST_MATH_MAX_DIGITS10_HPP
+
+namespace boost {
+namespace multiprecision {
+namespace detail {
+
+template <unsigned digits>
+struct calc_max_digits10
+{
+#ifndef BOOST_NO_CXX11_CONSTEXPR
+   static constexpr unsigned max_digits_10(unsigned d)
+   {
+      //
+      // We need ceil(log10(2) * d) + 1 decimal places to
+      // guarantee round tripping, see: https://www.exploringbinary.com/number-of-digits-required-for-round-trip-conversions/
+      // and references therein.  Since log10(2) is irrational, then d*log10(2) will
+      // never be exactly an integer so we can replace by trunc(log10(2) * d) + 2
+      // and avoid the call to ceil:
+      //
+      return static_cast<unsigned>(0.301029995663981195213738894724493026768189881462108541310 * d) + 2;
```

Comment:
How do you force compiler to use all these digits in this log10(2) value? I had problems forcing constexpr to work with `double` alone.

Reply:
Ah, my bad, those extra digits are wasted - it's a double precision constant.  BTW I tried a reasonably exhaustive search comparing double precision calculation to 200-digit calculation and couldn't find a bit count value where they disagreed.  float precision is a different matter though.

Reply:
Hmm, I think you could leave these wasted digits. Maybe someday the compilers will start picking them up :)

Since that's a double constant, with 15 (up o 17 :) digits, it means that this calculation will be incorrect for binary digits>1e15. That corresponds to 301029995663981 decimal digits. I didn't use MPFR_301029995663981 yet.

> float precision is a different matter though.

This calculation agrees for `float`, it has 24 bits and 6 digits10. And:

`trunc(log10(2)*24)+2 == 9`, just like in that table and on page 46 of IEEE standard.

> BTW I tried a reasonably exhaustive search comparing double precision calculation to 200-digit calculation and couldn't find a bit count value where they disagreed.

oh. you meant a kind of calculation which I talked a lot about in the earlier post? I had plenty of examples where they disagree. But now that I increased extraStringDigits10, I have a hard time finding them again :) Today is too late. I will check tomorrow. Thank you for trying this!

Reply:
> That corresponds to 301029995663981 decimal digits
... provided that the `unsigned` type is large enough :)

About these errors: maybe they way in which I calculate the relative error is to blame. I will check this tomorrow.

Reply:
Do you know a clean method of checking a particular bit value in (any) floating point significand?

Comparing bits would be much better, I suspect that my calculation of relative error has its own numerical inaccuracy problems :/

Reply:
boost::math::float_distance ?

Only for two types the same though.  For two different types, I would round the high precision type to the same number of bits as the low precision type so the comparison is against a "correctly rounded" value.

Reply:
I found a way with `frexp` function:

```
#include <boost/math/constants/constants.hpp>
#include <boost/multiprecision/float128.hpp>
#include <boost/multiprecision/mpfr.hpp>
#include <iostream>
#include <limits>
#include <vector>

template <typename T> int sign(T val) { return (T(0) < val) - (val < T(0)); }

class DecomposedReal {
private:
	int                        sig;
	int                        exp;
	std::vector<unsigned char> bits;

public:
	template <typename Rr> DecomposedReal(Rr x)
	{
		int ex   = 0;
		Rr  norm = frexp(abs(x), &ex);
		sig      = sign(x);
		exp      = ex - 1;
		ex       = 0;
		int pos  = 0;
		bits.resize(std::numeric_limits<Rr>::digits, 0);
		while (norm != 0) {
			assert((ex <= 0) and (pos < int(bits.size())) and (pos >= 0));
			pos -= ex;
			bits[pos] = 1;
			norm -= static_cast<Rr>(0.5);
			norm = frexp(norm, &ex);
		};
	}
	template <typename Rr> Rr rebuild()
	{
		Rr  ret = 0;
		int i   = 0;
		for (auto c : bits) {
			if (c != 0) {
				ret += pow(static_cast<Rr>(2), static_cast<Rr>(exp - i));
			}
			++i;
		}
		return ret * static_cast<Rr>(sig);
	}
	template <typename Rr = double> void print()
	{
		std::cout << "sign : " << sig << "\n";
		std::cout << "exp  : " << exp << "\n";
		std::cout << "bits : ";
		for (auto c : bits)
			std::cout << int(c);
		std::cout << "\nreconstructed number: " << rebuild<Rr>() << "\n\n";
	}
};

template <typename Rr> void test(const Rr& arg)
{
	std::cout << std::setprecision(std::numeric_limits<Rr>::digits10 + 3);
	std::cout << "original number     = " << arg << "\n";
	DecomposedReal d(arg);
	d.print<Rr>();
	assert(arg == d.rebuild<Rr>());
};

int main()
{
	test<float>(-1);                                  // 2^0
	test<double>(0);                                  // 0
	test<long double>(-4);                            // -2^2
	test<boost::multiprecision::float128>(0.375);     // 2^-2+2^-3
	test<boost::multiprecision::mpfr_float_50>(1036); // 2^3+2^2+2^10
	test<boost::multiprecision::mpfr_float_100>(boost::math::constants::pi<boost::multiprecision::mpfr_float_100>());
}
```

gives output:

```
original number     = -1
sign : -1
exp  : 0
bits : 100000000000000000000000
reconstructed number: -1

original number     = 0
sign : 0
exp  : -1
bits : 00000000000000000000000000000000000000000000000000000
reconstructed number: 0

original number     = -4
sign : -1
exp  : 2
bits : 1000000000000000000000000000000000000000000000000000000000000000
reconstructed number: -4

original number     = 0.375
sign : 1
exp  : -2
bits : 11000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
reconstructed number: 0.375

original number     = 1036
sign : 1
exp  : 10
bits : 100000011000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
reconstructed number: 1036

original number     = 3.141592653589793238462643383279502884197169399375105820974944592307816406286208998628034825342117067951
sign : 1
exp  : 1
bits : 1100100100001111110110101010001000100001011010001100001000110100110001001100011001100010100010111000000011011100000111001101000100101001000000100100111000001000100010100110011111001100011101000000001000001011101111101010011000111011000100111001101100100010010100010100101000001000011110011000111000110100000001001101110111101111100101
reconstructed number: 3.141592653589793238462643383279502884197169399375105820974944592307816406286208998628034825342117067951
```

I suppose that now I can do some bit comparison stuff :)